core/src/main/scala/org/http4s/rho/bits/EntityParser.scala

```diff
@@ -0,0 +1,49 @@
+package org.http4s
+package rho.bits
+
+import scala.language.higherKinds
+
+import scala.annotation.tailrec
+import scala.collection.generic.CanBuildFrom
+import org.http4s.EntityDecoder
+import org.http4s.DecodeFailure
+import scalaz.concurrent.Task
+
+/** Extract a value from the `Request` `Query`
+  *
+  * @tparam A Type of value produced by the parser.
+  */
+ trait EntityParser[A] { self => 
```

Comment:
I'm not sure about special treatment of an inner type like that.... How about how http4s does it using `handle`?  

The final product would look something like

```scala
GET / "foo" / "bar" ^ EntityDecoder[Baz] |>> { 
  ... 
} handleDecodeErrors { 
  ... 
}
```

However, I think this won't work with the existing architecture, because of the way routes are composed using some side-effects in the background.